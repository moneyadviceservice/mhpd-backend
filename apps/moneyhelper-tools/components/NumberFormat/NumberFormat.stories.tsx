import { StoryFn } from '@storybook/react';
import NumberFormat, { Props } from '.';

const StoryProps = {
  title: 'Components/TOOLS/NumberFormat',
  component: NumberFormat,
};

const Template: StoryFn<Props> = (args) => <NumberFormat {...args} />;

export const Default = Template.bind({});
Default.args = {
  value: 1000,
};

export const DirectionIncrease = Template.bind({});
DirectionIncrease.args = {
  value: 1000,
  direction: 'increase',
};

export const DirectionIncreaseWithIconRight = Template.bind({});
DirectionIncreaseWithIconRight.args = {
  value: 1000,
  direction: 'increase',
  position: 'right',
};

export const DirectionDecrease = Template.bind({});
DirectionDecrease.args = {
  value: 1000,
  direction: 'decrease',
};

export const DirectionDecreaseWithIconRight = Template.bind({});
DirectionDecreaseWithIconRight.args = {
  value: 1000,
  direction: 'decrease',
  position: 'right',
};

export default StoryProps;
