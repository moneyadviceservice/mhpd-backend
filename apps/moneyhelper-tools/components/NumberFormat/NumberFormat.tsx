import { NumericFormat, NumericFormatProps } from 'react-number-format';
import { twMerge } from 'tailwind-merge';

export type Props = NumericFormatProps & {
  direction?: 'increase' | 'decrease';
  position?: 'left' | 'right';
};

const NumberFormat = ({
  className,
  direction,
  position = 'left',
  ...props
}: Props) => {
  const arrowCSS = 'inline-block border-solid border-x-transparent';
  const isIncrease = direction === 'increase';
  const isDecrease = direction === 'decrease';

  const iconLeftOrder = 'flex-row-reverse';
  const iconRightOrder = 'flex-row';

  const getNumberWithIcon = (order: string) => {
    return (
      <span className={`flex items-center gap-2 ${order}`}>
        <NumericFormat
          className={className}
          decimalScale={2}
          fixedDecimalScale={true}
          displayType="text"
          thousandSeparator={true}
          {...props}
        />
        <span
          className={twMerge(
            arrowCSS,
            isIncrease &&
              'border-b-[13px] border-x-[8px] border-t-0 border-b-red-700',
            isDecrease &&
              'border-t-[13px] border-x-[8px] border-b-0 border-t-green-700',
          )}
        ></span>
      </span>
    );
  };

  if (direction && position === 'left') {
    return (
      <span className="inline-flex">{getNumberWithIcon(iconLeftOrder)}</span>
    );
  }

  if (direction && position === 'right') {
    return (
      <span className="inline-flex">{getNumberWithIcon(iconRightOrder)}</span>
    );
  }

  return (
    <NumericFormat
      className={className}
      decimalScale={2}
      displayType="text"
      thousandSeparator={true}
      {...props}
    />
  );
};

export default NumberFormat;
