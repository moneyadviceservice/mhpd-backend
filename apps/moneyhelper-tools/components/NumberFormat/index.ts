import NumberFormat from './NumberFormat';
export type { Props } from './NumberFormat';
export default NumberFormat;
