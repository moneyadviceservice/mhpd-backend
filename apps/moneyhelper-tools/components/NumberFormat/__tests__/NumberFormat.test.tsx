import { render } from '@testing-library/react';
import NumberFormat from '../NumberFormat';

describe('NumberFormat component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <NumberFormat id="id" name="name" defaultValue="" onChange={() => {}} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with a custom className', () => {
    const { container } = render(
      <NumberFormat
        className="rounded-lg"
        id="id"
        name="name"
        defaultValue=""
        onChange={() => {}}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with direction and position props', () => {
    const { container } = render(
      <NumberFormat
        id="id"
        name="name"
        defaultValue=""
        onChange={() => {}}
        direction="decrease"
        position="right"
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
