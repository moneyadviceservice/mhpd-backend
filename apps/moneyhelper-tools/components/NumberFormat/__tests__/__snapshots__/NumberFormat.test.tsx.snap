// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberFormat component renders correctly 1`] = `
<span
  id="id"
  name="name"
/>
`;

exports[`NumberFormat component renders correctly with a custom className 1`] = `
<span
  class="rounded-lg"
  id="id"
  name="name"
/>
`;

exports[`NumberFormat component renders correctly with direction and position props 1`] = `
<span
  class="inline-flex"
>
  <span
    class="flex items-center gap-2 flex-row"
  >
    <span
      id="id"
      name="name"
    />
    <span
      class="inline-block border-solid border-x-transparent border-t-[13px] border-x-[8px] border-b-0 border-t-green-700"
    />
  </span>
</span>
`;
